pkg/controllers/management/clusterprovisioner/helm_cleaner.go

```diff
@@ -0,0 +1,296 @@
+		var apiVersion string
+
+		if mappingData.DeprecatedInVersion != "" {
+			apiVersion = mappingData.DeprecatedInVersion
+		} else {
+			apiVersion = mappingData.RemovedInVersion
+		}
+
+		if !semver.IsValid(kubeVersion) {
+			return replaced, "", errors.Errorf("Invalid format for Kubernetes semantic version: %v", kubeVersion)
+		}
+
+		if !semver.IsValid(apiVersion) {
+			return replaced, "", errors.Errorf("Failed to get the deprecated or removed Kubernetes version for API: %s", strings.ReplaceAll(deprecatedAPI, "\n", " "))
+		}
+
+		if count := strings.Count(manifest, deprecatedAPI); count > 0 {
+			if semver.Compare(apiVersion, kubeVersion) > 0 {
+				logrus.Debugf("The following API does not require mapping as the "+
+					"API is not deprecated or removed in Kubernetes '%s':\n\"%s\"\n", apiVersion,
+					deprecatedAPI)
+			} else {
+				if supportedAPI == "" {
+					logrus.Debugf("Found %d instances of deprecated or removed Kubernetes API:\n\"%s\"\n", count, deprecatedAPI)
+
+					for repl := 0; repl < count; repl++ {
+						// find the position where the API header is
+						apiIndex := strings.Index(manifest, deprecatedAPI)
+
+						// find the next separator index
+						separatorIndex := strings.Index(manifest[apiIndex:], "---\n")
+
+						// find the previous separator index
+						previousSeparatorIndex := strings.LastIndex(manifest[:apiIndex], "---\n")
+
+						/*
+						 * if no previous separator index was found, it means the resource is at the beginning and not
+						 * prefixed by ---
+						 */
+						if previousSeparatorIndex == -1 {
+							if apiIndex == 0 {
+								previousSeparatorIndex = 0
+							} else {
+								previousSeparatorIndex = apiIndex - 1
+							}
+						}
+
+						if separatorIndex == -1 { // this means we reached the end of input
+							manifest = manifest[:previousSeparatorIndex]
+						} else {
+							manifest = manifest[:previousSeparatorIndex] + manifest[separatorIndex+apiIndex:]
+						}
+					}
+
+					manifest = strings.Trim(manifest, "\n")
+				} else {
+					logrus.Debugf("Found %d instances of deprecated or removed Kubernetes API:\n\"%s\"\nSupported API equivalent:\n\"%s\"\n", count, deprecatedAPI, supportedAPI)
+					manifest = strings.ReplaceAll(manifest, deprecatedAPI, supportedAPI)
+				}
+
+				replaced = true
+			}
+		}
+	}
+	return replaced, manifest, nil
+}
+
+// updateRelease updates a release in the cluster with an equivalent with the superseded APIs replaced or removed as
+// needed.
+// Logic extracted from https://github.com/helm/helm-mapkubeapis/blob/main/pkg/v3/release.go#L71-L94
+func updateRelease(originalRelease *release.Release, modifiedManifest string, config *action.Configuration) error {
+	originalRelease.Info.Status = release.StatusSuperseded
+	if err := config.Releases.Update(originalRelease); err != nil {
+		return errors.Wrapf(err, "[updateRelease] failed to update original release %v in namespace %v", originalRelease.Name, originalRelease.Namespace)
+	}
```

Comment:
Should we wait to do this or at least add logic to revert it if the subsequent release fails to create? Or is this practice?